FirebaseCore/Tests/Unit/FIRHeartbeatLoggerTest.m

```diff
@@ -54,6 +54,110 @@ - (void)tearDown {
   [FIRHeartbeatLoggingTestUtils removeUnderlyingHeartbeatStorageContainersAndReturnError:nil];
 }
 
+#pragma mark - Instance Management
+
+- (void)testGettingInstance_WithSameAppID_ReturnsSameInstance {
+  // Given
+  FIRHeartbeatLogger *heartbeatLogger1 = [FIRHeartbeatLogger loggerForAppID:@"appID"];
+  // When
+  FIRHeartbeatLogger *heartbeatLogger2 = [FIRHeartbeatLogger loggerForAppID:@"appID"];
+  // Then
+  NSLog(@"");
+  XCTAssertNotNil(heartbeatLogger1);
+  XCTAssertNotNil(heartbeatLogger2);
+  XCTAssert(heartbeatLogger1 == heartbeatLogger2, "Instances should reference the same object.");
+
+  __auto_type __weak weakHeartbeatLogger1 = heartbeatLogger1;
+  __auto_type __weak weakHeartbeatLogger2 = heartbeatLogger2;
+  [self addTeardownBlock:^{
+    XCTAssertNil(weakHeartbeatLogger1);
+    XCTAssertNil(weakHeartbeatLogger2);
+  }];
+}
+
+- (void)testGettingInstance_WithDifferentAppID_ReturnsDifferentInstances {
+  // Given
+  FIRHeartbeatLogger *heartbeatLogger1 = [FIRHeartbeatLogger loggerForAppID:@"appID_1"];
+  // When
+  FIRHeartbeatLogger *heartbeatLogger2 = [FIRHeartbeatLogger loggerForAppID:@"appID_2"];
+  // Then
+  XCTAssertNotNil(heartbeatLogger1);
+  XCTAssertNotNil(heartbeatLogger2);
+  XCTAssert(heartbeatLogger1 != heartbeatLogger2, "Instances should reference the same object.");
+
+  __auto_type __weak weakHeartbeatLogger1 = heartbeatLogger1;
+  __auto_type __weak weakHeartbeatLogger2 = heartbeatLogger2;
+  [self addTeardownBlock:^{
+    XCTAssertNil(weakHeartbeatLogger1);
+    XCTAssertNil(weakHeartbeatLogger2);
+  }];
+}
+
+- (void)testCachedInstancesCannotBeRetainedWeakly {
+  // Given
+  FIRHeartbeatLogger *strongHeartbeatLogger = [FIRHeartbeatLogger loggerForAppID:@"appID"];
+  FIRHeartbeatLogger *__weak weakHeartbeatLogger = nil;
+  @autoreleasepool {
+    weakHeartbeatLogger = [FIRHeartbeatLogger loggerForAppID:@"appID"];
+  }
+  XCTAssertNotNil(strongHeartbeatLogger);
+  XCTAssertNotNil(weakHeartbeatLogger);
+  XCTAssert(strongHeartbeatLogger == weakHeartbeatLogger,
+            "Instances should reference the same object.");
+  // When
+  strongHeartbeatLogger = nil;
+  // Then
+  XCTAssertNil(strongHeartbeatLogger);
+  XCTAssertNil(weakHeartbeatLogger);
+}
+
+- (void)testCachedInstancesAreRemovedUponDeinitAndCanBeRetainedStrongly {
```

Comment:
RemovedUponDeinit? is that a typo?

Reply:
It's not a typo, but I should make it clearer to readers.

Cached instances are stored at the _class_ level of `FIRHeartbeatLogger` in a dictionary of `NSString` appIDs mapping to weakly boxed `FIRHeartbeatLogger`. When a `FIRHeartbeatLogger`'s retain count goes to 0, I want it to be removed from the `cachedInstances`. IIUC, this should also prevent the (rare but possible?) growing of `FIRHeartbeatLogger` instances stored if the end user kept configuring different Firebase apps.

In the `dealloc` of `FIRHeartbeatLogger`, it removes itself from the `cachedInstances`:
https://github.com/firebase/firebase-ios-sdk/blob/722ec3e3b400df6695f16ac3ba3c48eb950970f2/FirebaseCore/Sources/FIRHeartbeatLogger.m#L103-L106

In this test method, I want to verify that cached instance are removed (from the cachedInstances static storage) when they are dealloc'd (`heartbeatLogger1` and `heartbeatLogger2`) but can be strongly retained (`heartbeatLogger3`).

I'll change to:
```suggestion
- (void)testCachedInstancesAreRemovedUponDeallocButCanBeRetainedStrongly {
```

Lmk if this sounds reasonable!

Reply:
Reworked in 666f6e1b3a9406904178f3902cb233f444c698e8